### ISSUE TITLE
Switch to Authorization Code Grant Flow for authentication

### ISSUE THREAD

**Issue by demize** [fluid-queue/fluid-queue#17]
Right now, the bot uses Implicit Grant Flow, usually with a forever token from twitchapps. This works, but forever tokens are deprecated, and implicit tokens going forward have a max 60-day validity. To address this, we should switch to the Authorization Code Grant flow prior to our first official release.

This will mean a few things need to be implemented:

- [ ] A helper tool to complete the authorization process
- [x] Code within the bot to validate the token, refresh it if necessary, and connect to IRC with it
- [x] Hourly token validation performed by the bot (This is a [requirement](https://dev.twitch.tv/docs/authentication/validate-tokens) from Twitch)

We *could* also implement a toggle between implicit/auth code, but I would recommend against that; we don't know when forever will end, and I'd rather not encourage users to find out on their own.

**Comment by HvDaedalus** [comment]
Suggestion for a helper tooling setup:

Expressjs (since we already have a node application here) to have a lightweight webserver for simple and easy operations.

What expressjs would do: Provide an API for the reactjs app to save/update the configuration in the bot and serve the reactjs app.

What the reactjs app would do: Provide a UI helping the user authenticate with twitch and provision the configuration for the bot. The reactjs app development should not be part of the bot application, so it needs its own repository. The distributable content of the reactjs app however will be part of the bot application.

**Comment by demize** [comment]
We'd probably want to submodule the react app, then. Would keep it out of the dev tree, but still make it accessible within the repository (so long as you remember to `git submodule update --init`).

**Comment by HvDaedalus** [comment]
That's not really what you want to do here. Submoduling would make the source of the reactjs app part of the bot. What we actually want is the distributed (released) version of the app. We could have the bot check for newer versions on startup and provide the newest one, if needed.

**Comment by demize** [comment]
Ah yeah, I think we basically need to do it that way.

On one hand, it'd be fine if we kept `main` stable in the react app, because then the submodule would always get the most up to date version (though it would still have the source). But that would also require distributing the bot through git, or uploading our own archives with each release, and both those options are less than great.

We don't necessarily need to have the bot pull it down itself, though. The React app is only really needed once (or if your refresh token expires), so it can be better handled through setup instructions:

```console
$ wget https://github.com/fluid-queue/react-helper/releases/latest/download/react-helper.tar.gz
$ tar xvf react-helper.tar.gz && rm react-helper.tar.gz
```

is all you'd really need, though we'd want to provide a Windows version of the instructions as well. That said, there'd definitely be a benefit to doing it in the bot, to allow for automatic updates in case tokens expire.

If we do have the bot pull it down, I think we'd want the flow to look something like this:

1. Bot starts up, checks its current token
2. If that token's expired, refresh it
3. If the refresh token is expired, or if there's no token configuration, update the React app and prompt the user to finish configuration.

That avoids needing to check for an update every boot, and lets the user delete the React app if they don't need it anymore without it coming back on its own. We will need a way to check the latest version, though; I think that's pretty easy to accomplish with a static site, just disguised as an API. I can get that set up easily enough on Github Pages, we'd just need to make sure to update it as part of our post-release checklists.

**Comment by demize** [comment]
Merged as part of 2.0.0-rc.1.
